cfme/tests/webui/test_general_ui.py

```diff
@@ -523,7 +527,10 @@ def test_provider_details_page_refresh_after_clear_cookies(
 
     # When the test runs a second time for cloud provider, it raises an error,
     # this finalizer is workaround for it.
-    request.addfinalizer(appliance.server.logout)
+    @request.addfinalizer
+    def _finalize():
+        appliance.server.login()
+        appliance.server.logout()
```

Comment:
Why do we login and then logout at the end of the test?

Reply:
This is to make sure the next parametrized test runs without any error. This is a little difficult to explain. So if we don't use the finalizer to login and logout, this is what happens - 
1. You are on Infra provider details page, the session timeouts, you click on any button on the infra details page and you are redirected to the login page.
(If you login with this redirected login page, you will be directed to the same page you were, before getting logged out.)
2. The first test ran successfully, so infra provider is removed and the cloud provider is added.
3. You use the same login screen from step 1 to login and navigate to the cloud provider's details page, you are directed to infra provider's details page, but since there is no infra provider, you see a flash error and an exception is raised by `is_displayed`.

If we login and logout, we will be directed to the dashboard(or whatever start page is) and then we can navigate to the cloud provider's details page. Does this make sense?

Reply:
Ah, I see, thanks for the explanation. So in step 3) immediately after login, the user is redirected to the infra provider details page since this is the last page on which their session was active? Why can't we just navigate away from that page?

Reply:
> So in step 3) immediately after login, the user is redirected to the infra provider details page since this is the last page on which their session was active?


Correct.

>Why can't we just navigate away from that page?

I'm not sure if I understand the question correctly. Are you suggesting we navigate to some other page as a part of finalizer instead of re-login?

Reply:
Q: is not best way to just `logout`?